### DISCUSSION TITLE
SA 1.4 alternative for __primary_entity__ from SA 1.3

### DISCUSSION THREAD

**Discussion by kmhosny** [sqlalchemy/sqlalchemy#10559] [Usage Questions] [answered]
Hi There 👋 ,

I'm working on a codebase which uses flask with sqlalchemy 1.3, I'm trying to do an upgrade to 1.4 but since this code is somehow old it heavily relies on 1.3. 
The code uses a class CustomQuery which inherits from sqlalchemy.orm.Query class, in this class it heavily relies on an attribute called __primary_entity__ which seems to have been removed in 1.4. What's the use of this attribute and what could be the replacement for it?
Here is the snippet that uses it:
```python
if query._primary_entity is None:  # pylint: disable=protected-access
        return query
base_class = query._primary_entity.mapper.class_  # pylint: disable=protected-access
```

**Comment by zzzeek** [comment, marked as answer]
it's likely trying to give you what the main "class" is, that is if you had a query like this:

    q = session.query(User)

the primary entity would be `User`.

in all modern SQLAlchemy versions going back for years, there is public API for this information at [Query.column_descriptions](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.column_descriptions).   The zeroth element would contain the primary entity, if any:

    q = session.query(User)
    assert q.column_descriptions[0]['entity'] is User

**Reply by kmhosny** [reply]
that's makes sense to use. thanks @zzzeek !
